server/events.go

```diff
@@ -0,0 +1,294 @@
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+const (
+	connectEventSubj    = "$SYS.%s.CLIENT.CONNECT"
+	disconnectEventSubj = "$SYS.%s.CLIENT.DISCONNECT"
+)
+
+// ConnectEventMsg is sent when a new connection is made that is part of an account.
+type ConnectEventMsg struct {
+	Server ServerInfo `json:"server"`
+	Client ClientInfo `json:"client"`
+}
+
+// DisconnectEventMsg is sent when a new connection previously defined from a
+// ConnectEventMsg is closed.
+type DisconnectEventMsg struct {
+	Server   ServerInfo `json:"server"`
+	Client   ClientInfo `json:"client"`
+	Sent     DataStats  `json:"sent"`
+	Received DataStats  `json:"received"`
+	Reason   string     `json:"reason"`
+}
+
+type ServerInfo struct {
+	Host string `json:"host"`
+	ID   string `json:"id"`
+	Seq  uint64 `json:"seq"`
+}
+
+// ClientInfo is detailed information about the client forming a connection.
+type ClientInfo struct {
+	Start   time.Time  `json:"start,omitempty"`
+	Host    string     `json:"host,omitempty"`
+	ID      uint64     `json:"id"`
+	Account string     `json:"acc"`
+	User    string     `json:"user,omitempty"`
+	Name    string     `json:"name,omitempty"`
+	Lang    string     `json:"lang,omitempty"`
+	Version string     `json:"ver,omitempty"`
+	Stop    *time.Time `json:"stop,omitempty"`
+}
+
+// DataStats reports how may msg and bytes. Applicable for both sent and received.
+type DataStats struct {
+	Msgs  int64 `json:"msgs"`
+	Bytes int64 `json:"bytes"`
+}
+
+// Used for internally queueing up messages that the server wants to send.
+type pubMsg struct {
+	r   *SublistResult
+	sub string
+	msg []byte
+}
+
+func (s *Server) internalSendLoop() {
+	defer s.grWG.Done()
+	s.mu.Lock()
+	if s.sys == nil {
+		return
+	}
+	c := s.sys.client
+	acc := s.sys.account
+	sendq := s.sys.sendq
+	s.mu.Unlock()
+
+	for s.isRunning() {
+		select {
+		case pm := <-sendq:
+			// Prep internal structures needed to send message.
+			c.pa.subject = []byte(pm.sub)
+			c.pa.size = len(pm.msg)
+			c.pa.szb = []byte(strconv.FormatInt(int64(len(pm.msg)), 10))
+			// Add in NL
+			pm.msg = append(pm.msg, _CRLF_...)
+			// Check to see if we need to map/route to another account.
+			if acc.imports.services != nil {
+				c.checkForImportServices(acc, pm.msg)
+			}
+			c.processMsgResults(acc, pm.r, pm.msg, []byte(pm.sub), nil, nil)
+			c.flushClients()
+		case <-s.quitCh:
+			return
+		}
+	}
+}
+
+// This will queue up a message to be sent.
+func (s *Server) sendInternalMsg(r *SublistResult, sub string, msg []byte) {
+	if s.sys == nil {
+		return
+	}
+	s.sys.sendq <- &pubMsg{r, sub, msg}
```

Comment:
Should we use a select? Are we ok if we lose those messages?
As long as we know for a fact that there won't be any locking causing the "sendLoop" to be blocked by the code calling sendInternalMsg() when chan buffer is full, I guess we are ok, otherwise, use combination of buffered channel and select, or move to list.

Reply:
Its a buffered channel now.

Reply:
yes, I saw it was a buffered of 128, but say that there is rapid generation of events that fill the buffered channel, then the sendloop is accessing a lock that another routine is holding trying to send to the channel. that's what I meant. If we know for sure that this situation will never occur, then we are fine.

Reply:
We don't hold any locks when placing on the sendq or pulling off.

Reply:
Did a change though to make more consistent.